dali/operators/image/color/brightness_contrast.cc

```diff
@@ -114,31 +118,43 @@ bool BrightnessContrastCpu::SetupImpl(std::vector<OutputDesc> &output_desc,
-                    kernels::KernelContext ctx;
-                    auto tvin = view<const InputType, 3>(input[sample_id]);
-                    auto tvout = view<OutputType, 3>(output[sample_id]);
-                    float add, mul;
-                    OpArgsToKernelArgs<OutputType, InputType>(add, mul,
-                      brightness_[sample_id], brightness_shift_[sample_id], contrast_[sample_id]);
-                    kernel_manager_.Run<Kernel>(thread_id, sample_id, ctx, tvout, tvin,
-                                                add, mul);
-                }, out_shape.tensor_size(sample_id));
-              }
-          }
-      ), DALI_FAIL(make_string("Unsupported output type: ", output_type_)))  // NOLINT
-  ), DALI_FAIL(make_string("Unsupported input type: ", input.type())))  // NOLINT
-  tp.RunAll();
+    const auto &input = ws.template Input<CPUBackend>(0);
+    auto &output = ws.template Output<CPUBackend>(0);
+    output.SetLayout(input.GetLayout());
+    auto out_shape = output.shape();
+    auto& tp = ws.GetThreadPool();
+    TYPE_SWITCH(input.type(), type2id, InputType, (uint8_t, int16_t, int32_t, float), (
+        TYPE_SWITCH(output_type_, type2id, OutputType, (uint8_t, int16_t, int32_t, float), (
+            {
+                using Kernel = TheKernel<OutputType, InputType>;
+                TensorListShape<> sh = input.shape();
+                auto num_dims = sh.sample_dim();
+                int frame_num = num_dims == 3 ? 1 : sh[0][0];
+                for (int sample_id = 0; sample_id < input.shape().num_samples(); sample_id++) {
+                    for (int frame_id; frame_id < frame_num; frame_id++) {
+                        auto sample_shape = out_shape.tensor_shape_span(sample_id);
+                        auto vol = volume(sample_shape.begin() + static_cast<int>(num_dims != 3),
+                                          sample_shape.end());
+                        tp.AddWork([&, sample_id, frame_id, num_dims](int thread_id) {
+                            kernels::KernelContext ctx;
+                            auto tvin = num_dims == 3 ? view<const InputType, 3>(input[sample_id]) :
+                                subtensor(view<const InputType, 4>(input[sample_id]), frame_id);
+                            auto tvout = num_dims == 3 ? view<OutputType, 3>(output[sample_id]) :
+                                subtensor(view<OutputType, 4>(output[sample_id]), frame_id);
+                            float add, mul;
+                            OpArgsToKernelArgs<OutputType, InputType>(add, mul,
+                                                                      brightness_[sample_id],
+                                                                      brightness_shift_[sample_id],
+                                                                      contrast_[sample_id]);
+                            kernel_manager_.Run<Kernel>(thread_id, sample_id, ctx, tvout, tvin,
```

Comment:
Technically, this is wrong. Now we've multipled the number of samples, so it's `sum(num_frames[i] for i in range(num_samples)`. Since the kernel is stateless, we can stop pretending that there are multiple samples and just pass 0 as sample_id all the time. It will be less confusing.

Reply:
Done